acceptance/builder_test.go

```diff
@@ -0,0 +1,184 @@
+package acceptance
+
+import (
+	"fmt"
+	"math/rand"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/sclevine/spec"
+	"github.com/sclevine/spec/report"
+
+	"github.com/buildpacks/lifecycle/api"
+	h "github.com/buildpacks/lifecycle/testhelpers"
+)
+
+var (
+	builderDockerContext = filepath.Join("testdata", "builder")
+	builderBinaryDir     = filepath.Join("testdata", "builder", "container", "cnb", "lifecycle")
+	builderImage         = "lifecycle/acceptance/builder"
+	builderUserID        = "1234"
+)
+
+func TestBuilder(t *testing.T) {
+	h.SkipIf(t, runtime.GOOS == "windows", "builder acceptance tests are not yet supported on Windows")
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	//h.MakeAndCopyLifecycle(t, "linux", builderBinaryDir)
+	h.DockerBuild(t,
+		builderImage,
+		builderDockerContext,
+		h.WithArgs("--build-arg", fmt.Sprintf("cnb_platform_api=%s", api.Platform.Latest())),
+	)
+	//defer h.DockerImageRemove(t, builderImage)
+
+	spec.Run(t, "acceptance-builder", testBuilder, spec.Parallel(), spec.Report(report.Terminal{}))
+}
+
+func testBuilder(t *testing.T, when spec.G, it spec.S) {
+	// .../cmd/lifecycle/builder.go#45
+	when("called with arguments", func() {
+		it("errors", func() {
+			command := exec.Command(
+				"docker",
+				"run",
+				"--rm",
+				"--env", "CNB_PLATFORM_API="+latestPlatformAPI,
+				builderImage,
+				"some-arg",
+			)
+			output, err := command.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to parse arguments: received unexpected arguments"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+
+	// .../cmd/lifecycle/builder.go#62
+	when("running as a root", func() {
+		it("errors", func() {
+			command := exec.Command(
+				"docker",
+				"run",
+				"--rm",
+				"--user",
+				"root",
+				"--env", "CNB_PLATFORM_API="+latestPlatformAPI,
+				builderImage,
+			)
+			output, err := command.CombinedOutput()
+			h.AssertNotNil(t, err)
+			expected := "failed to build: refusing to run as root"
+			h.AssertStringContains(t, string(output), expected)
+		})
+	})
+	when("error on reading Data", func() {
+
+		// .../cmd/lifecycle/builder.go#116-118
+		when("read buildpack group file", func() {
+			it("no default group toml file in default location", func() {
+				command := exec.Command(
+					"docker",
+					"run",
+					"--entrypoint", "/cnb/delete_file_then_run.sh",
+					"--rm",
+					"--env", "CNB_PLATFORM_API="+latestPlatformAPI,
+					"--env", "DELETE_FILE=/layers/group.toml",
```

Comment:
This setup strikes me as a bit convoluted - I wonder if we could pass the group location via the `-group` arg.

Reply:
(and then when no file exists at that location, you can make the same assertion)

Reply:
Yes it make sense, but also with this way i think we are testing these lines too, am i right ? Because it is also testing default path location.

https://github.com/buildpacks/lifecycle/blob/1d7a850cf03a1e16b6be5f871419d41091da4f29/cmd/lifecycle/builder.go#L48-L50

If u want i can add one more test case for non-exist file test or i can change with that, which one make sense for u ?

Reply:
Actually here is how it's done for detector tests: https://github.com/buildpacks/lifecycle/blob/771ac0b9d0d56cd3baaf28355498a0a636ea6182/acceptance/detector_test.go#L87

So the builder image wouldn't have a `/layers/group.toml` but it could have a `/layers/other-group.toml` that is provided with the `-group` flag. There would be two tests - one to show a failure when no `-group` is passed and one to show success when `-group /layers/other-group.toml` is passed.

Reply:
If I understood correctly, I will delete the `/layers/group.toml` file directly. I will not put group.toml in the default search location. Instead I will create `cnb/group_tomls/always_detect_group.toml` and use that for success scenario. Also for all other test cases I need to pass this file like `CNB_GROUP_PATH=cnb/group_tomls/always_detect_group.toml`.

Reply:
I did some changes according to my latest comment. Please check